README.md
# flutter_transition_animation
![](https://badgen.net/github/license/micromatch/micromatch)
![](https://img.shields.io/pub/v/flutter_transition_animation)

🔥🔥🔥可配置的动画转场，用来展示顶部提示信息更香哦

## 效果🔥🔥🔥
![](flutter_transform_msg.gif)

## 如何使用
#### pub
在工程文件`yaml`写入下面配置：

```dart

flutter_transition_animation: ^*.*.*

```
> ⚠️ [最新版本查看](https://pub.dev/packages/flutter_transition_animation/install)
>
> ![](https://img.shields.io/pub/v/flutter_transition_animation)

#### 导入

```
import 'package:flutter_transition_animation/flutter_easy_message.dart';

```

## 🐎动画步骤

动画进入可以定制进入的时间，动画规则，和动画的方向。

|动画方向|含义|
|:-:|:-:|
|rtl| 右到左|
|ltr|左到右|
|ttb| 顶部到底部|
|btt|底部到顶部|
|thv|水平合并到一条线|
|tvc|竖直合成到一条线|

|动画规则|含义|
|:-:|:-:|
|easy| 渐隐渐显|
|easIn| 缓慢进入|
|line|匀速|


> 更多规则见: [animation](https://api.flutter.dev/flutter/animation/Curves-class.html)

|时间属性|含义|
|:-:|:-:|
|duration| 动画停留时间|
|showDuration| 动画从没有到完全展示时间|
|hideDuration| 动画消失使用的时间|


## 🚗 执行动画


> ⚠️ 只需要设定不一样的key即可。


## 🌰🌰🌰examaple

```dart

Container(
  height: 200,
  width: 300,
  color: Colors.black12,
  /// 每次刷新给不同 key 可以执行动画
  key: UniqueKey(),
  alignment: Alignment.topCenter,
  child: Container(
    width: 200,
    height: 250,
    child: FlutterEasyMessage(
      child: child,
      messageChild: messageChild,
      duration: Duration(seconds: 2),
      hideAnimationStyle: dismissStyle,
      showAnimationStyle: style,
      showAnimaitonDirection: showFlipStyle,
      hideAnimaitonDirection: hideFlipStyle,
      hideDuration: Duration(seconds: 1),
      showDuration: Duration(seconds: 1),
      curve: _showCurve,
      reverseCurve: _hideCurve,
    ),
  ),
)
```

> [查看完整例子]( ./example)


## 详细属性

```dart
/// 主体 部件
  final Widget child;

  /// 展示 动画的部件
  final Widget messageChild;

  /// 动画停留时间展示时间默认 [1000ms]
  final Duration duration;

  /// 展示动画时间 默认 [300ms]
  final Duration showDuration;

  /// 隐藏动画时间 默认 [300ms]
  final Duration hideDuration;

  /// 动画出现类别 默认是[EasyShowOrHideStyle.defaultStyle]
  final EasyShowOrHideStyle showAnimationStyle;

  /// 动画消失类别 默认是[EasyShowOrHideStyle.defaultStyle]
  final EasyShowOrHideStyle hideAnimationStyle;

  /// 动画出现时候的方向，默认是[EasyAnimaitonDirection.ltr]
  final EasyAnimationDirection showAnimaitonDirection;

  /// 动画隐藏时候的方向,默认是[EasyAnimaitonDirection.btt]
  final EasyAnimationDirection hideAnimaitonDirection;

  /// 进入的时候的规则,默认是[Curves.linear] [https://api.flutter.dev/flutter/animation/Curves-class.html]
  final Curve curve;

  /// 返回时候 规则,默认是[Curves.linear][https://api.flutter.dev/flutter/animation/Curves-class.html]
  final Curve reverseCurve;
```


## ❤️❤️❤️[喜欢就留下你的✨✨](https://github.com/ifgyong/flutter_transition_animation)

## 🔐[License](LICENSE)

CHANGELOG.md
## [1.0.0]
 > 1.0 上传包


lib/easy_message.dart
import 'dart:async';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'flutter_easy_message.dart';

///
/// Created by fgyong on 2020/9/10.
///

enum EasyShowOrHideStyle {
  defaultStyle,
  /*渐隐渐显*/
  fade,
  /*大小变化*/
  size,
  /*翻转*/
  flip,
}

// ignore: must_be_immutable
class FlutterEasyMessage extends StatelessWidget {
  /// 主体 部件
  final Widget child;

  /// 展示 动画的部件
  final Widget messageChild;

  /// 动画停留时间展示时间默认 [1000ms]
  final Duration duration;

  /// 展示动画时间 默认 [300ms]
  final Duration showDuration;

  /// 隐藏动画时间 默认 [300ms]
  final Duration hideDuration;

  /// 动画出现类别 默认是[EasyShowOrHideStyle.defaultStyle]
  final EasyShowOrHideStyle showAnimationStyle;

  /// 动画消失类别 默认是[EasyShowOrHideStyle.defaultStyle]
  final EasyShowOrHideStyle hideAnimationStyle;

  /// 动画出现时候的方向，默认是[EasyAnimaitonDirection.ltr]
  final EasyAnimationDirection showAnimaitonDirection;

  /// 动画隐藏时候的方向,默认是[EasyAnimaitonDirection.btt]
  final EasyAnimationDirection hideAnimaitonDirection;

  /// 进入的时候的规则,默认是[Curves.linear] [https://api.flutter.dev/flutter/animation/Curves-class.html]
  final Curve curve;

  /// 返回时候 规则,默认是[Curves.linear][https://api.flutter.dev/flutter/animation/Curves-class.html]
  final Curve reverseCurve;

  FlutterEasyMessage({
    @required this.child,
    @required this.messageChild,
    @required this.duration,
    this.showDuration,
    this.hideDuration,
    this.curve = Curves.linear,
    this.reverseCurve = Curves.linear,
    this.showAnimationStyle,
    this.hideAnimationStyle,
    this.showAnimaitonDirection = EasyAnimationDirection.ltr,
    this.hideAnimaitonDirection = EasyAnimationDirection.btt,
  });

  @override
  Widget build(BuildContext context) {
    return _DefaultWidget(
      child: child,
      messageChild: messageChild,
      duration: duration,
      showAnimation: getAnimation(showAnimationStyle),
      hideAnimation: getAnimation(hideAnimationStyle),
      showDuration: showDuration,
      hideAnimationDirection: hideAnimaitonDirection,
      showAnimationDirection: showAnimaitonDirection,
      reverseCurve: reverseCurve,
      curve: curve,
    );
  }

  EasyAnimaiton defaultAnimation = EasyFadeAnimaiton();

  /// 根据类型获取到处理动画的类
  EasyAnimaiton getAnimation(EasyShowOrHideStyle animationStyle) {
    switch (animationStyle) {
      case EasyShowOrHideStyle.fade:
        return defaultAnimation;
      case EasyShowOrHideStyle.size:
        return EasySizeAnimaiton();
      case EasyShowOrHideStyle.flip:
        return EasyFlipAnimaiton();
      default:
        return defaultAnimation;
    }
  }
}

class _DefaultWidget extends StatefulWidget {
  final Widget child;
  final Widget messageChild;
  final Duration duration;
  final Curve curve;

  final EasyAnimaiton showAnimation;
  final EasyAnimaiton hideAnimation;

  /// 动画出现时候的方向
  final EasyAnimationDirection showAnimationDirection;

  /// 动画隐藏时候的方向
  final EasyAnimationDirection hideAnimationDirection;

  /// 展示动画时间 默认 300ms
  final Duration showDuration;

  /// 隐藏动画时间 默认 300ms
  final Duration hideDuration;

  /// 返回时候 规则
  /// 查看所有效果 [https://api.flutter.dev/flutter/animation/Curves-class.html]
  final Curve reverseCurve;

  /// 默认的 提示框
  _DefaultWidget(
      {@required this.child,
      @required this.messageChild,
      @required this.duration,
      this.curve,
      this.showDuration,
      this.hideDuration,
      @required this.showAnimation,
      this.hideAnimation,
      this.showAnimationDirection,
      this.hideAnimationDirection,
      this.reverseCurve})
      : assert(child != null),
        assert(messageChild != null),
        assert(duration != null),
        assert(
          showAnimation != null,
        );

  @override
  __DefaultWidgetState createState() {
    return __DefaultWidgetState();
  }
}

class __DefaultWidgetState extends State<_DefaultWidget> {
  Widget get child => widget.child;
  Widget get messageChild => widget.messageChild;

  /// 停留的周期
  Duration get duration => widget.duration;
  bool _dismiss = false;

  /// 出现周期
  Duration get showDuration =>
      widget.showDuration ?? Duration(milliseconds: 300);

  /// 隐藏 周期
  Duration get hideDuration =>
      widget.hideDuration ?? Duration(milliseconds: 300);

  /// 出现动画
  EasyAnimaiton get easyAnimaiton => widget.showAnimation;

  /// 隐藏动画
  EasyAnimaiton get easyHideAnimaiton =>
      widget.hideAnimation ?? widget.showAnimation;

  @override
  Widget build(BuildContext context) {
    return Container(
      key: ValueKey(_dismiss),
      child: EasyDismiss(
        builder: (animation, context) {
          Widget ch;
          if (_dismiss == true) {
            ch = easyHideAnimaiton.animationWidget(
                animation, context, messageChild,
                style: widget.hideAnimationDirection, show: false);
          } else {
            ch = easyAnimaiton.animationWidget(animation, context, messageChild,
                style: widget.showAnimationDirection, show: true);
          }
          // return ch;
          return Stack(
            children: [
              Positioned(
                child: child,
              ),
              Positioned(
                child: Container(
                  child: ch,
                ),
              )
            ],
          );
        },
        duration: showDuration,
        reverseDuration: hideDuration,
        reverse: _dismiss,
        reverseCurve: widget.reverseCurve,
        curve: widget.curve,
      ),
    );
  }

  @override
  void initState() {
    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      /// 隐藏小部件
      Timer(widget.duration, () {
        _dismiss = true;
        if (mounted) setState(() {});
      });
    });
    super.initState();
  }

  @override
  void dispose() {
    super.dispose();
  }
}


lib/animation/flip.dart
import 'dart:math';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import '../flutter_easy_message.dart';
import 'animation.dart';
import 'fade.dart';

///
/// Created by fgyong on 2020/9/10.
///

class EasyFlipAnimaiton extends EasyDirectionAnimation {
  @override
  Widget animationWidget(
      Animation<double> animation, BuildContext context, Widget child,
      {EasyAnimationDirection style, bool show}) {
    Alignment alignment;

    /// 根据枚举类型来指定不同的锚点进行动画
    switch (style) {
      case EasyAnimationDirection.ttb:
        if (show == false) {
          alignment = Alignment.bottomCenter;
        } else {
          alignment = Alignment.topCenter;
        }
        break;

      case EasyAnimationDirection.btt:
        alignment =
            show == false ? Alignment.topCenter : Alignment.bottomCenter;
        break;

      case EasyAnimationDirection.ltr:
        alignment =
            show == false ? Alignment.centerRight : Alignment.centerLeft;
        break;

      case EasyAnimationDirection.rtl:
        alignment = show == true ? Alignment.centerRight : Alignment.centerLeft;
        break;

      case EasyAnimationDirection.thv:
      case EasyAnimationDirection.tvc:
        alignment = Alignment.center;
        break;
    }
    return Transform(
      transform: m4(animation.value, style),
      child: child,
      alignment: alignment,
    );
    // return ;
  }

  Matrix4 m4(double value, EasyAnimationDirection style) {
    var m = Matrix4.identity()..setEntry(3, 2, 0.001);

    switch (style) {
      case EasyAnimationDirection.ttb:
      case EasyAnimationDirection.btt:
      case EasyAnimationDirection.thv:

        /// 解决向上翻 90°没有完全隐藏问题
        if (value == 0) {
          m..setEntry(3, 2, 0.0);
        }
        return m..rotateX((1 - value) * pi / 2);
      case EasyAnimationDirection.rtl:
      case EasyAnimationDirection.ltr:
      case EasyAnimationDirection.tvc:

        /// 解决向上翻 90°没有完全隐藏问题
        if (value == 0) {
          m..setEntry(3, 2, 0.0);
        }
        return m..rotateY((1 - value) * pi / 2);
    }
    return m;
  }
}


lib/animation/size.dart
import 'dart:math';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_transition_animation/flutter_easy_message.dart';

///
/// Created by fgyong on 2020/9/11.
///
class EasySizeAnimaiton extends EasyAnimaiton {
  @override
  Widget animationWidget(
      Animation<double> animation, BuildContext context, Widget child,
      {EasyAnimationDirection style, bool show}) {
    Alignment alignment = Alignment.center;

    /// 根据枚举类型来指定不同的锚点进行动画
    switch (style) {
      case EasyAnimationDirection.ttb:
        if (show == false) {
          alignment = Alignment.bottomCenter;
        } else {
          alignment = Alignment.topCenter;
        }
        break;

      case EasyAnimationDirection.btt:
        alignment =
            show == false ? Alignment.topCenter : Alignment.bottomCenter;
        break;

      case EasyAnimationDirection.ltr:
        alignment =
            show == false ? Alignment.centerRight : Alignment.centerLeft;
        break;

      case EasyAnimationDirection.rtl:
        alignment = show == true ? Alignment.centerRight : Alignment.centerLeft;
        break;

      case EasyAnimationDirection.thv:
      case EasyAnimationDirection.tvc:
        alignment = Alignment.center;
        break;
    }
    return Transform(
      transform: m4(animation.value, style),
      child: child,
      alignment: alignment,
    );
    // return ;
  }

  Matrix4 m4(double value, EasyAnimationDirection style) {
    var m = Matrix4.identity();

    switch (style) {
      case EasyAnimationDirection.ttb:
      case EasyAnimationDirection.btt:
      case EasyAnimationDirection.thv:
        return m
          ..scale(
            1.0,
            value,
          );
      case EasyAnimationDirection.rtl:
      case EasyAnimationDirection.ltr:
      case EasyAnimationDirection.tvc:
        return m..scale(value, 1.0);
    }
    return m;
  }
}


lib/dismissController/dismiss_controller.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

///
/// Created by fgyong on 2020/9/10.
///

typedef DismissAcionCallBack = Widget Function(
    Animation<double>, BuildContext context);

class EasyDismiss extends StatefulWidget {
  final Widget child;
  final Curve curve;

  /// 返回时候 规则
  final Curve reverseCurve;
  final DismissAcionCallBack builder;
  final double upperBound, lowerBound;
  final Duration reverseDuration, duration;

  /// 是否是反向
  final bool reverse;

  /// 执行一次的动画 部件
  EasyDismiss(
      {Key key,
      this.child,
      this.curve = Curves.linear,
      @required this.builder,
      this.lowerBound = 0.0,
      this.upperBound = 1.0,
      this.reverse,
      this.duration,
      this.reverseDuration,
      this.reverseCurve})
      : assert(() {
          if (lowerBound != null &&
              upperBound != null &&
              lowerBound < upperBound) return true;
          return false;
        }()),
        super(key: key);
  @override
  State<StatefulWidget> createState() => __DismissState();
}

class __DismissState extends State<EasyDismiss> with TickerProviderStateMixin {
  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      builder: (context, child) {
        return widget.builder(
          _animation,
          context,
        );
      },
      animation: _animation,
    );
  }

  AnimationController _animationController;
  Animation<double> _animation;
  @override
  void initState() {
    _animationController = AnimationController(
      vsync: this,
      upperBound: widget.upperBound,
      reverseDuration: widget.reverseDuration,
      duration: widget.duration,
      lowerBound: widget.lowerBound,
    );
    if (widget.reverse == true) {
      _animationController.reverse(from: widget.upperBound);
    } else {
      _animationController.forward();
    }
    _animation = CurvedAnimation(parent: _animationController, curve: curve);
    super.initState();
  }

  Curve get reverseCurve => widget.reverseCurve ?? widget.curve;
  Curve get curve => widget.reverse == false ? widget.curve : reverseCurve;

  @override
  void dispose() {
    if (_animationController != null) {
      _animationController.stop(canceled: true);
      _animationController.dispose();
    }

    super.dispose();
  }
}


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_transition_animation/flutter_easy_message.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Easy show message Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

enum ChildType { msg, img, view }

class _MyHomePageState extends State<MyHomePage> {
  EasyAnimationDirection showFlipStyle = EasyAnimationDirection.btt;
  EasyAnimationDirection hideFlipStyle = EasyAnimationDirection.btt;

  EasyShowOrHideStyle style = EasyShowOrHideStyle.defaultStyle;
  EasyShowOrHideStyle dismissStyle = EasyShowOrHideStyle.defaultStyle;

  Curve _showCurve = Curves.linear;
  Curve _hideCurve = Curves.linear;
  ChildType type = ChildType.msg;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          // mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            SizedBox(
              height: 50,
            ),
            Container(
              height: 200,
              width: 300,
              color: Colors.black12,
              key: UniqueKey(),
              alignment: Alignment.topCenter,
              child: Container(
                width: 280,
                height: 250,
                child: FlutterEasyMessage(
                  child: child,
                  messageChild: messageChild,
                  duration: Duration(seconds: 2),
                  hideAnimationStyle: dismissStyle,
                  showAnimationStyle: style,
                  showAnimaitonDirection: showFlipStyle,
                  hideAnimaitonDirection: hideFlipStyle,
                  hideDuration: Duration(seconds: 1),
                  showDuration: Duration(seconds: 1),
                  curve: _showCurve,
                  reverseCurve: _hideCurve,
                ),
              ),
            ),
            Text('浮窗类型'),
            _widgetChild(),
            Text('出现动画类型'),
            _widgetStyle(true),
            Text('消失动画类型'),
            _widgetStyle(false),
            Text('隐藏方向  style==EasyShowOrHideStyle.size 可用 \n '),
            Text('出现动画方向'),
            _widget(true),
            Text('隐藏动画方向'),
            _widget(false),
            Text('进入动画曲线<左右滑动选择更多>'),
            CupertinoScrollbar(
              child: SingleChildScrollView(
                child: _curve(true),
                scrollDirection: Axis.horizontal,
              ),
            ),
            Text('隐藏动画曲线<左右滑动选择更多>'),
            CupertinoScrollbar(
              child: SingleChildScrollView(
                child: _curve(false),
                scrollDirection: Axis.horizontal,
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          setState(() {});
        },
        tooltip: 'Increment',
        child: Icon(Icons.refresh),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }

  void fresh() {
    if (mounted) setState(() {});
  }

  /// 动画类型
  Widget _widgetStyle(bool show) {
    return CupertinoSegmentedControl<EasyShowOrHideStyle>(
        children: {
          EasyShowOrHideStyle.defaultStyle: Padding(
            padding: EdgeInsets.all(5.0),
            child: Text("defaultStyle"),
          ),
          EasyShowOrHideStyle.size: Padding(
            padding: EdgeInsets.all(5.0),
            child: Text("size"),
          ),
          EasyShowOrHideStyle.fade: Padding(
            padding: EdgeInsets.all(5.0),
            child: Text("fade"),
          ),
          EasyShowOrHideStyle.flip: Padding(
            padding: EdgeInsets.all(5.0),
            child: Text("翻转"),
          )
        },
        selectedColor: Colors.blueAccent,
        onValueChanged: (v) {
          if (show == false) {
            dismissStyle = v;
          } else {
            style = v;
          }
          fresh();
        },
        groupValue: show == true ? style : dismissStyle);
  }

  Widget _widget(bool show) {
    return CupertinoSegmentedControl<EasyAnimationDirection>(
        children: {
          EasyAnimationDirection.rtl: Padding(
            padding: EdgeInsets.all(5.0),
            child: Text("<-"),
          ),
          EasyAnimationDirection.ltr: Padding(
            padding: EdgeInsets.all(5.0),
            child: Text("->"),
          ),
          EasyAnimationDirection.btt: Padding(
            padding: EdgeInsets.all(5.0),
            child: Text("↑"),
          ),
          EasyAnimationDirection.ttb: Padding(
            padding: EdgeInsets.all(5.0),
            child: Text("↓"),
          ),
          EasyAnimationDirection.tvc: Padding(
            padding: EdgeInsets.all(5.0),
            child: Text(show ? "<-->" : ">-<"),
          ),
          EasyAnimationDirection.thv: Padding(
            padding: EdgeInsets.all(5.0),
            child: Text(show ? "↕" : "--"),
          )
        },
        selectedColor: Colors.blueAccent,
        onValueChanged: (v) {
          if (show) {
            showFlipStyle = v;
          } else {
            hideFlipStyle = v;
          }

          fresh();
        },
        groupValue: show == true ? showFlipStyle : hideFlipStyle);
  }

  Widget _widgetChild() {
    return CupertinoSegmentedControl<ChildType>(
      children: {
        ChildType.msg: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text("msg"),
        ),
        ChildType.img: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text("img"),
        ),
        ChildType.view: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text("view"),
        ),
      },
      groupValue: type,
      onValueChanged: (v) {
        type = v;
        fresh();
      },
    );
  }

  Widget _curve(bool show) {
    return CupertinoSegmentedControl<Curve>(
      children: {
        Curves.linear: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('linear'),
        ),
        Curves.bounceInOut: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('bounceInOut'),
        ),
        Curves.bounceOut: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('bounceOut'),
        ),
        Curves.bounceIn: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('bounceIn'),
        ),
        Curves.easeIn: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('easeIn'),
        ),
        Curves.easeInToLinear: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('easeInToLinear'),
        ),
        Curves.linearToEaseOut: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('linearToEaseOut'),
        ),
        Curves.bounceIn: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('bounceIn'),
        ),
        Curves.decelerate: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('decelerate'),
        ),
        Curves.easeInBack: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('easeInBack'),
        ),
        Curves.easeInCirc: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('easeInCirc'),
        ),
        Curves.easeInCubic: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('easeInCubic'),
        ),
        Curves.easeInExpo: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('easeInExpo'),
        ),
        Curves.easeInOut: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('easeInOut'),
        ),
        Curves.easeInOutBack: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('easeInOutBack'),
        ),
        Curves.easeInOutCirc: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('easeInOutCirc'),
        ),
        Curves.slowMiddle: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('slowMiddle'),
        ),
        Curves.fastOutSlowIn: Padding(
          padding: EdgeInsets.all(5.0),
          child: Text('fastOutSlowIn'),
        ),
      },
      onValueChanged: (v) {
        if (show == true) {
          _showCurve = v;
        } else {
          _hideCurve = v;
        }
        fresh();
      },
      groupValue: show == true ? _showCurve : _hideCurve,
    );
  }

  // ignore: missing_return
  Widget get messageChild {
    switch (type) {
      case ChildType.msg:
        return Container(
          child: Text(
            '今天有大雨，记得带伞🌲😯',
          ),
          height: 30,
          alignment: Alignment.center,
          decoration: BoxDecoration(
              color: Colors.black38,
              borderRadius: BorderRadius.all(Radius.circular(10))),
        );
      case ChildType.img:
        return Container(
          child: Image.asset(
            'img/0.jpeg',
            fit: BoxFit.fitWidth,
          ),
          height: 100,
          width: 200,
          decoration: BoxDecoration(
              color: Colors.black38,
              borderRadius: BorderRadius.all(Radius.circular(10))),
        );
      case ChildType.view:
        return ClipRRect(
          borderRadius: BorderRadius.all(Radius.circular(10)),
          child: Container(
              child: Row(
                children: [
                  Image.asset(
                    'img/0.jpeg',
                    fit: BoxFit.fitWidth,
                  ),
                  SizedBox(
                    width: 50,
                  ),
                  Expanded(
                    child: Text(
                      'www.fgyong.cn \n 快来关注我 www.github/ifgyong',
                      style: TextStyle(color: Colors.white),
                    ),
                  )
                ],
              ),
              height: 80,
              // width: 200,
              color: Colors.black38),
        );
    }
  }

  Widget get child => Container(
        padding: EdgeInsets.only(top: 100),
        child: OutlineButton(
          child: Text('我是 child，可以 \n 点击 弹出提示框'),
          onPressed: () {
            setState(() {});
          },
        ),
      );
}
